app/models/theme_field.rb

```diff
@@ -277,6 +277,82 @@ def target_name
     Theme.targets.invert[target_id].to_s
   end
 
+  class ThemeFileMatcher
+    OPTIONS = %i{name type target}
+    def initialize(regex, canonical:, target:, name:, type:)
+      @allowed_values = {}
+      @allowed_values[:name] = Array(name) if name
+      @allowed_values[:target] = Array(target) if target
+      @allowed_values[:type] = Array(type) if type
+      @canonical = canonical
+      @regex = regex
+    end
+
+    def opts_from_filename(filename)
+      match = @regex.match(filename)
+      return false unless match
+      hash = {}
+      OPTIONS.each do |option|
+        hash[option] = @allowed_values[option][0] if @allowed_values[option].length == 1
+        hash[option] = match[option] if hash[option].nil?
+      end
+      hash
+    end
+
+    def filename_from_opts(opts)
+      is_match = true
+      OPTIONS.each do |option|
+        next if @allowed_values[option] == nil
```

Comment:
Could do this in one line if you prefer:

`next if (@allowed_values[option] || []).include?(opts[option])`

Reply:
The idea was that it should always match when the value is `nil`. But this was a very confusing piece of code, so I've refactored it to be clearer, and added a couple of comments.

Reply:
The `names`/`targets`/`types` values can only be set via the initialiser, where they are required keyword arguments. So it's impossible to leave them "unset".